### PR TITLE
Remove lodash dependency

### PR DESCRIPTION
**Pull request by tylershuster**
Tell me if this isn't welcome, but

I think it's twenty twenty and transpiled (Java/Type)Script can do a lot more than when lodash was invented. It's not like landscape leans heavily on it. There are a *few* places where the vanilla replacement is more awkward but I think in general native JavaScript is more expressive. It's a little bit lighter.

### CODE REVIEW

**Comment by matildepark** [comment]
@tacryt-socryp @liam-fitzgerald How do you feel about removing the lodash dependency? I believe you both made use of it; I'm pretty neutral on this.

**Comment by liam-fitzgerald** [comment]
:( 

in defense of lodash:

> I think it's twenty twenty and transpiled (Java/Type)Script can do a lot more than when lodash was invented.

This is unequivocally true. However, there are still huge holes in the JS stdlib. Map/reduce/filter over objects is still awkward, requiring `Object.keys` or `Object.entries`. Objects as a generic key value datastructure are pretty fundamental and I'd hate to lose these tools. `lodash/fp` is also neat for point free style. Furthermore, lodash's typings are *really* good and prevent you from having to reason about TS generics and avoid casts. e.g. lodash's `_.pick`, when correctly typed looks roughly like this. 
```typescript
type Pick<O, K extends keyof O> = { [X in keyof O]: X extends K ? O[X] : never };

function pick<O, K extends keyof O>(obj: O, keys: O[]): Pick<O,K>
```
These type signatures are awkward to read and write and should be avoided if we can.

Lodash is not particularly heavy, and so removing it feels kind of unnecessary. 

RFC though, interested to hear others opinions.

**Comment by liam-fitzgerald** [comment]
Screen Shot 2020-08-04 at 11 36 18 am

A visualisation of our deps. FWIW, they all gzip to 337KB so i don't think we have a particularly onerous dependency issue right now.

File: pkg/interface/src/apps/chat/components/lib/chat-input.js [hunk 1]
```diff
@@ -339,12 +338,10 @@ export class ChatInput extends Component {
...
 
-    const candidates = _.chain(this.props.envelopes)
-      .defaultTo([])
-      .map('author')
-      .uniq()
-      .reverse()
-      .value();
+    const candidates = this.props.envelopes
+      .map(envelope => envelope.author)
+      .filter((value, index, self) => self.indexOf(value) === index)
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/chat/components/lib/chat-input.js:343]: this is more verbose && probably less efficient than _.uniq

Comment (liam-fitzgerald) [reply pkg/interface/src/apps/chat/components/lib/chat-input.js:343]: also this doesn't handle the envelopes is undefined case

File: pkg/interface/src/apps/chat/components/lib/ship-search.js [hunk 2]
```diff
@@ -119,35 +118,30 @@ export class ShipSearch extends Component {
 
       return (
         hay.startsWith(needle) ||
-        _.some(_.words(hay), s => s.startsWith(needle))
+        hay.split(/\s/).some(s => s.startsWith(needle))
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/chat/components/lib/ship-search.js:121]: this is not technically equivalent, but it's close enough. Leaves in punctuation and numbers.

File: pkg/interface/src/apps/publish/app.js [hunk 3]
```diff
@@ -46,19 +45,17 @@ export default class PublishApp extends React.Component {
 
     const notebooks = props.notebooks ? props.notebooks : {};
 
-    const unreadTotal = _.chain(notebooks)
-      .values()
-      .map(_.values)
-      .flatten() // flatten into array of notebooks
+    const unreadTotal = Object.values(notebooks)
+      .map((owner) => Object.values(owner))
```
Comment (liam-fitzgerald) [inline pkg/interface/src/apps/publish/app.js:49]: Another bug bear: point free style is pretty difficult with JS stdlib.

File: pkg/interface/src/components/InviteSearch.tsx [hunk 4]
```diff
@@ -177,7 +178,7 @@ export class InviteSearch extends Component<
       const shipIdx = shipMatches.findIndex((ship) => ship === selected);
       const staleSelection = groupIdx < 0 && shipIdx < 0;
       if (!selected || staleSelection) {
-        const newSelection = _.get(groupMatches, '[0][0]') || shipMatches[0];
+        const newSelection = (groupMatches[0] && groupMatches[0].length) ? groupMatches[0][0] : shipMatches[0];
```
Comment (liam-fitzgerald) [inline pkg/interface/src/components/InviteSearch.tsx:181]: should be able to use the elvis operator here instead

**Comment by tylershuster** [comment]
Seems like a cowboy hunting dependencies might set his sights on codemirror instead 🤔

T

File: pkg/interface/src/lib/util.js [hunk 5]
```diff
@@ -88,7 +86,7 @@ function hexToDec(hex) {
 }
 
 export function hexToRgba(hex, a) {
-  const [r,g,b] = _.chunk(hex, 2).map(hexToDec);
+  const [r,g,b] = /^([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex).map(hexToDec);
```
Comment (liam-fitzgerald) [inline pkg/interface/src/lib/util.js:89]: probably want to include uppercase here.

**Comment by ohAitch** [comment]
But syntax highlighting! …I wonder how big a hacked-down read-only build of
codemirror would be

**Comment by tylershuster** [comment]
FWIW, I don’t particularly have a strong opinion about removing lodash. Just going for low hanging fruit. But if it’s not ripe I won’t pull

T

**Comment by tylershuster** [comment]
@liam-fitzgerald your comments are all well-taken. I think some of my edits are fine, but not really enough for a PR. I would love to see Landscape done purely functionally but I don't think it's the time for that even if it is the Right Thing to Do. Unless @tacryt-socryp has some really crushing arguments in favor of removing lodash, I'm going to close this and say thank you for your consideration.

**Pull request closed without merging**